data_pipeline_api/file_api.py

```diff
@@ -1,91 +1,162 @@
+        )
+
+    @staticmethod
+    def construct_metadata_store(metadata_filename: Path):
+        """Construct a MetadataStore corresponding to a file.
+        """
+        if metadata_filename.exists():
+            logger.debug("loading metadata from %s", metadata_filename)
+            with open(metadata_filename) as metadata_store_file:
+                metadata_store = yaml.safe_load(metadata_store_file)
+        else:
+            logger.warning("could not find metadata.yaml")
+            metadata_store = {}
+        return MetadataStore(metadata_store)
+
+    def __init__(self, config_filename: Union[Path, str]):
+        """The FileAPI class provides tracked interaction with the filesystem.
+
+        Files are located through a metadata lookup system based on two files: a
+        metadata database file, which lists all known files and their associated
+        metadata, and a configuration file, which provides mechanisms for influencing
+        the metadata lookup process.
+        """
+        logger.debug("config_filename = %s", config_filename)
         self._open_timestamp = datetime.now()
+        logger.debug("open_timestamp = %s", self._open_timestamp)
         self._accesses = []
         config_filename = Path(config_filename)
+        config_root = config_filename.parent
 
         try:
             with open(config_filename) as config_file:
                 self._config = yaml.safe_load(config_file)
                 if "run_id" in self._config:
                     self.run_id = self._config["run_id"]
+                    logger.debug("run_id = %s (from config)", self.run_id)
                 else:
                     # Compute a run_id.
-                    with open(config_filename, "rb") as file:
-                        m = sha1(file.read())
-                    m.update(str(self._open_timestamp).encode())
-                    self.run_id = m.hexdigest()
+                    self.run_id = FileAPI.calculate_hash(
+                        config_filename, str(self._open_timestamp).encode()
+                    )
+                    logger.debug("run_id = %s (computed)", self.run_id)
+
                 self.data_directory = Path(self._config.get("data_directory", "."))
+                logger.debug("data_directory = %s", self.data_directory)
+
                 self.fail_on_hash_mismatch = self._config.get(
                     "fail_on_hash_mismatch", True
                 )
-                self._read_overrides = Overrides(
-                    (override.get("where", {}), override.get("use", {}))
-                    for override in self._config.get("read", ())
+                logger.log(
+                    DEBUG if self.fail_on_hash_mismatch else WARNING,
+                    "fail_on_hash_mismatch = %s",
+                    self.fail_on_hash_mismatch,
+                )
+
+                self._read_overrides = FileAPI.construct_overrides(
+                    self._config.get("read", ())
```

Comment:
should these be the AccessType constants?

Reply:
Hmm, quite possibly, though it is arguable - the link is in some sense only a coincidence, because these are just the names of config sections and don't actually propagate through to the accesses directly. Additionally, this is now in a tower of four PRs, so I'd rather not change it 🤷 .